File: server/services/progressiveQueue.ts
import { CharacterEntry, QuestionSetup, Theme } from '../types';
import {
  buildFootballPrimaryHint,
  getFootballHintFields,
} from './footballHints';
import { getNBAExtraHintFields } from './nbaHints';
import { getPokemonBonusHintFields } from './pokemonHints';
import { getHintFields } from './dataLoader';
import { isHintFieldExcluded } from './footballHints';
import { isNBAHintFieldExcluded } from './nbaHints';

function shuffle<T>(arr: T[], rng?: { next(): number }): T[] {
  const copy = [...arr];
  const rand = (n: number) => (rng ? Math.floor(rng.next() * n) : Math.floor(Math.random() * n));
  for (let i = copy.length - 1; i > 0; i--) {
    const j = rand(i + 1);
    [copy[i], copy[j]] = [copy[j], copy[i]];
  }
  return copy;
}

export function buildProgressiveHintQueueFromSetup(
  theme: Theme,
  setup: QuestionSetup,
  answer: CharacterEntry,
  rng?: { next(): number }
): {
  firstField: string;
  pendingQueue: string[];
  footballPrimaryField?: 'confederation' | 'clubLeague';
} {
  if (theme === 'football') {
    const primary = buildFootballPrimaryHint(answer, rng);
    const pool = shuffle(
      getFootballHintFields().filter((f) => f !== primary.field),
      rng
    );
    return {
      firstField: primary.field,
      pendingQueue: pool,
      footballPrimaryField: primary.field as 'confederation' | 'clubLeague',
    };
  }

  if (theme === 'nba') {
    const pool = shuffle(getNBAExtraHintFields(), rng);
    return {
      firstField: 'divisionPosition',
      pendingQueue: pool.filter((f) => f !== 'divisionPosition'),
    };
  }

  if (theme === 'pokemon') {
    const bonus = getPokemonBonusHintFields(setup.activeFields).filter(
      (f) => f !== setup.hintField
    );
    const fromActive = setup.activeFields.filter(
      (f) => f !== setup.hintField && !bonus.includes(f)
    );
    const pool = [...new Set([...fromActive, ...bonus, ...setup.extraHintFields])].filter(
      (f) => f !== setup.hintField
    );
    return {
      firstField: setup.hintField,
      pendingQueue: shuffle(pool, rng),
    };
  }

  if (theme === 'anime') {
    const pool = getHintFields(theme, setup.activeFields).filter(
      (f) => f !== 'name' && f !== 'anime' && f !== 'affiliation' && f !== setup.hintField
    );
    return {
      firstField: setup.hintField,
      pendingQueue: shuffle(pool, rng),
    };
  }

  const pool = getHintFields(theme, setup.activeFields).filter(
    (f) =>
      f !== 'name' &&
      f !== setup.hintField &&
      !isHintFieldExcluded(f) &&
      !isNBAHintFieldExcluded(f)
  );
  return {
    firstField: setup.hintField,
    pendingQueue: shuffle(pool, rng),
  };
}

File: server-go/internal/services/progressive_queue.go
package services

import (
	"guess-who/server-go/internal/types"
)

// BuildProgressiveHintQueueFromSetup 从题目配置构建首条提示字段与待解锁队列。
func BuildProgressiveHintQueueFromSetup(
	theme types.Theme,
	setup types.QuestionSetup,
	answer types.CharacterEntry,
	rng *SeededRng,
) (firstField string, pendingQueue []string, footballPrimaryField *string) {
	switch theme {
	case types.ThemeFootball:
		primary := BuildFootballPrimaryHint(answer, rng, "")
		pool := rngShuffleStrings(
			filterStrings(GetFootballHintFields(), func(f string) bool { return f != primary.Field }),
			rng,
		)
		fp := primary.Field
		return primary.Field, pool, &fp

	case types.ThemeNBA:
		pool := rngShuffleStrings(GetNBAExtraHintFields(), rng)
		filtered := filterStrings(pool, func(f string) bool { return f != "divisionPosition" })
		return "divisionPosition", filtered, nil

	case types.ThemePokemon:
		bonus := filterStrings(GetPokemonBonusHintFields(setup.ActiveFields), func(f string) bool {
			return f != setup.HintField
		})
		var fromActive []string
		for _, f := range setup.ActiveFields {
			if f != setup.HintField && !containsString(bonus, f) {
				fromActive = append(fromActive, f)
			}
		}
		poolSet := uniqueStrings(append(append(fromActive, bonus...), setup.ExtraHintFields...))
		pool := filterStrings(poolSet, func(f string) bool { return f != setup.HintField })
		return setup.HintField, rngShuffleStrings(pool, rng), nil

	default:
		pool := filterStrings(GetHintFields(theme, setup.ActiveFields), func(f string) bool {
			return f != "name" && f != setup.HintField && !IsHintFieldExcluded(f) && !IsNBAHintFieldExcluded(f)
		})
		return setup.HintField, rngShuffleStrings(pool, rng), nil
	}
}

func rngShuffleStrings(arr []string, rng *SeededRng) []string {
	if rng == nil {
		rng = NewSeededRng(1)
	}
	return rng.ShuffleStrings(arr)
}

func filterStrings(arr []string, keep func(string) bool) []string {
	out := make([]string, 0, len(arr))
	for _, s := range arr {
		if keep(s) {
			out = append(out, s)
		}
	}
	return out
}

func uniqueStrings(arr []string) []string {
	seen := make(map[string]bool)
	var out []string
	for _, s := range arr {
		if seen[s] {
			continue
		}
		seen[s] = true
		out = append(out, s)
	}
	return out
}

func containsString(arr []string, s string) bool {
	for _, v := range arr {
		if v == s {
			return true
		}
	}
	return false
}
